src/wallet/wallet.cpp

```diff
@@ -221,30 +221,36 @@ static std::set<std::string> g_unloading_wallet_set GUARDED_BY(g_wallet_release_
 // Custom deleter for shared_ptr<CWallet>.
 static void ReleaseWallet(CWallet* wallet)
 {
+    LOCK(g_wallet_release_mutex);
     const std::string name = wallet->GetName();
     wallet->WalletLogPrintf("Releasing wallet\n");
     wallet->Flush();
     delete wallet;
     // Wallet is now released, notify UnloadWallet, if any.
-    {
-        LOCK(g_wallet_release_mutex);
-        if (g_unloading_wallet_set.erase(name) == 0) {
-            // UnloadWallet was not called for this wallet, all done.
-            return;
-        }
+    if (g_unloading_wallet_set.erase(name) == 0) {
+        // UnloadWallet was not called for this wallet, all done.
+        return;
     }
     g_wallet_release_cv.notify_all();
 }
 
-void UnloadWallet(std::shared_ptr<CWallet>&& wallet)
+bool UnloadWallet(std::shared_ptr<CWallet>&& wallet)
 {
     // Mark wallet for unloading.
+    bool wallet_already_marked_for_unloading = false;
     const std::string name = wallet->GetName();
     {
         LOCK(g_wallet_release_mutex);
         auto it = g_unloading_wallet_set.insert(name);
-        assert(it.second);
```

Comment:
If this would be the issue presented in #29073, the test error logs would be showing a node crash. It wouldn't reach the `got_loading_error` check.

Reply:
Correct, this was not the issue as initially thought. The removal of this is only so that we return an appropriate message if it ever comes to this instead of crashing the node. I can change it back if it matters. Not sure if it makes sense this to lead to a node crash to be honest.

Reply:
> The removal of this is only so that we return an appropriate message if it ever comes to this instead of crashing the node. I can change it back if it matters. Not sure if it makes sense this to lead to a node crash to be honest.

The current structure of the code disallows it from happening; prior to every `UnloadWallet()` call, we call `RemoveWallet()`. Which ensures that we never crash at this point (https://github.com/bitcoin/bitcoin/pull/29143#issuecomment-1869817009).
So, if it crashes, the assertion failure makes sense because it means we are pushing more than one pointer of the same wallet into the `WalletContext` wallets vector.
We could document this behavior but, as is a different topic, it should be treated separately. Either on a standalone commit or a separate PR.

Also, agree with Marko that further explanations about the failure cause would be desirable. e.g. why moving the lock up to the top of `ReleaseWallet` fixes the problematic?